src/pool/srv_pool.c

```diff
@@ -5303,6 +5303,101 @@ ds_pool_extend_handler(crt_rpc_t *rpc)
 	crt_reply_send(rpc);
 }
 
+static int
+pool_discard(crt_context_t ctx, struct pool_svc *svc, struct pool_target_addr_list *list)
+{
+	struct pool_tgt_discard_in	*ptdi_in;
+	struct pool_tgt_discard_out	*ptdi_out;
+	struct rdb_tx			tx;
+	struct pool_map			*map;
+	crt_rpc_t			*rpc;
+	d_rank_list_t			*rank_list = NULL;
+	crt_opcode_t			opc;
+	int				i;
+	int				nr = 0;
+	int				rc;
+
+	rc = rdb_tx_begin(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term, &tx);
+	if (rc != 0)
+		return rc;
+
+	ABT_rwlock_wrlock(svc->ps_lock);
+	/* Create a temporary pool map based on the last committed version. */
+	rc = read_map(&tx, &svc->ps_root, &map);
+	if (rc != 0)
+		D_GOTO(out_lock, rc);
+
+	rank_list = d_rank_list_alloc(list->pta_number);
+	if (rank_list == NULL)
+		D_GOTO(out_map, rc = -DER_NOMEM);
+
+	/* remove the duplicate ranks from list, see reintegrate target case */
+	for (i = 0; i < list->pta_number; i++) {
+		struct pool_domain	*dom;
+		bool			exist = false;
+		int			j;
+
+		for (j = 0; j < nr; j++) {
+			if (rank_list->rl_ranks[j] == list->pta_addrs[i].pta_rank) {
+				exist = true;
+				break;
+			}
+		}
+		if (exist)
+			continue;
+
+		dom = pool_map_find_node_by_rank(map, list->pta_addrs[i].pta_rank);
+		if (dom->do_comp.co_status != PO_COMP_ST_DOWNOUT) {
+			D_DEBUG(DB_MD, DF_UUID": skip discard rank %u status %u\n",
+				DP_UUID(svc->ps_pool->sp_uuid), list->pta_addrs[i].pta_rank,
+				dom->do_comp.co_status);
+			continue;
+		}
+
+		rank_list->rl_ranks[nr++] = list->pta_addrs[i].pta_rank;
+		D_DEBUG(DB_MD, DF_UUID": discard rank %u\n",
+			DP_UUID(svc->ps_pool->sp_uuid), list->pta_addrs[i].pta_rank);
+	}
+
+	if (nr == 0) {
+		D_DEBUG(DB_MD, DF_UUID" discard 0 rank.\n", DP_UUID(svc->ps_pool->sp_uuid));
+		D_GOTO(out, rc = 0);
+	}
+
+	rank_list->rl_nr = nr;
+	opc = DAOS_RPC_OPCODE(POOL_TGT_DISCARD, DAOS_POOL_MODULE, DAOS_POOL_VERSION);
+	rc = crt_corpc_req_create(ctx, NULL, rank_list, opc, NULL,
+				  NULL, CRT_RPC_FLAG_FILTER_INVERT,
+				  crt_tree_topo(CRT_TREE_KNOMIAL, 32), &rpc);
```

Comment:
minor, 32 probably is too large, for most cases it is just flat tree. maybe 4 or 8 is enough.